krm-functions/nad-fn/mutator/mutator.go

```diff
@@ -0,0 +1,190 @@
+			},
+			Watch: map[corev1.ObjectReference]condkptsdk.WatchCallbackFn{
+				{
+					APIVersion: infrav1alpha1.GroupVersion.Identifier(),
+					Kind:       reflect.TypeOf(infrav1alpha1.ClusterContext{}).Name(),
+				}: m.ClusterContextCallbackFn,
+				{
+					APIVersion: ipamv1alpha1.GroupVersion.Identifier(),
+					Kind:       ipamv1alpha1.IPAllocationKind,
+				}: nil,
+				{
+					APIVersion: vlanv1alpha1.GroupVersion.Identifier(),
+					Kind:       vlanv1alpha1.VLANAllocationKind,
+				}: nil,
+				{
+					APIVersion: nephioreqv1alpha1.GroupVersion.Identifier(),
+					Kind:       nephioreqv1alpha1.InterfaceKind,
+				}: nil,
+			},
+			PopulateOwnResourcesFn: nil,
+			GenerateResourceFn:     m.generateResourceFn,
+		},
+	)
+	if err != nil {
+		rl.Results = append(rl.Results, fn.ErrorResult(err))
+	}
+	return m.fnCondSdk.Run()
+}
+
+func (r *mutatorCtx) ClusterContextCallbackFn(o *fn.KubeObject) error {
+	clusterContext := clusterctxtlibv1alpha1.NewMutator(o.String())
+	cluster, err := clusterContext.UnMarshal()
+	if err != nil {
+		return err
+	}
+	if cluster.Spec.CNIConfig.MasterInterface == "" {
+		return fmt.Errorf("MasterInterface on ClusterContext cannot be empty")
+	} else {
+		r.masterInterface = cluster.Spec.CNIConfig.MasterInterface
+	}
+	if cluster.Spec.CNIConfig.CNIType == "" {
+		return fmt.Errorf("CNIType on ClusterContext cannot be empty")
+	} else {
+		r.cniType = cluster.Spec.CNIConfig.CNIType
+	}
+	if cluster.Spec.SiteCode == nil {
+		return fmt.Errorf("SiteCode on ClusterContext cannot be empty")
+	} else {
+		r.siteCode = *cluster.Spec.SiteCode
+	}
+	r.clusterContextSet = true
+	return nil
+}
+
+func (r *mutatorCtx) generateResourceFn(forObj *fn.KubeObject, objs fn.KubeObjects) (*fn.KubeObject, error) {
+	// verify all needed objects exist
+	if objs.Where(fn.IsGroupVersionKind(nephioreqv1alpha1.InterfaceGroupVersionKind)).Len() == 0 {
+		return nil, fmt.Errorf("expected %s object to generate the nad", nephioreqv1alpha1.InterfaceKind)
+	}
+	if objs.Where(fn.IsGroupVersionKind(ipamv1alpha1.IPAllocationGroupVersionKind)).Len() == 0 &&
+		objs.Where(fn.IsGroupVersionKind(vlanv1alpha1.VLANAllocationGroupVersionKind)).Len() == 0 {
+		return nil, fmt.Errorf("expected one of %s or %s objects to generate the nad", ipamv1alpha1.IPAllocationKind, vlanv1alpha1.VLANAllocationKind)
+	}
+	if !r.clusterContextSet {
+		return nil, fmt.Errorf("expected ClusterContext object to generate the nad")
+	}
+
+	// generate an empty nad struct
+	nad, err := nadlibv1.NewFromGoStruct(&nadv1.NetworkAttachmentDefinition{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: nadv1.SchemeGroupVersion.Identifier(),
+			Kind:       reflect.TypeOf(nadv1.NetworkAttachmentDefinition{}).Name(),
+		},
+		ObjectMeta: metav1.ObjectMeta{Name: objs[0].GetName()},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	interfaces := objs.Where(fn.IsGroupVersionKind(nephioreqv1alpha1.InterfaceGroupVersionKind))
```

Comment:
This will be nitpicking from my part, sorry in  advance :) 
You are calling `objs.Where(fn.IsGroupVersionKind(nephioreqv1alpha1.InterfaceGroupVersionKind))` twice: once here, once in the beginning of the function. This can be avoided by assigning the result of the first run to the `interfaces` var.